test/run-unit-tests.py

```diff
@@ -33,9 +32,11 @@ def argument_parser():
 
 opts = argument_parser().parse_args()
 
-tests = glob.glob('/usr/lib/systemd/tests/test-*')
+unittestdir = pathlib.Path(__file__).parent.absolute()
```

Comment:
Oh, OK. Now I get it: it finds the tests from the temporary installation directory, while previously it found tests under `/usr/lib`, and probably there were no tests there and the glob was empty, or there were some installed ones from the previously-built package. Right now it fails because `/tmp` is mounted noexec, or there's some some MAC or whatever.

So this actually exposes a bug: `run-unit-tests.py` cannot be executed from the build directory, because various tests call helpers under real installed paths. E.g. `test-fstab-generator.sh` calls either `$1` or `/usr/lib/systemd/system-generators/systemd-fstab-generator`. When executed through `meson test`, we provide an argument so that `$1` points to `build/systemd-fstab-generator`. But when it is executed through `run-unit-tests.py`, there is no provision for this, so it just executes using the installation path. On my system, this fails, because `systemd-fstab-generator` changed recently and the installed version expects a slightly different output.

So… whoever is calling `/tmp/autopkgtest.1bn0Ov/build.uKj/systemd/test/run-unit-tests.py` needs to be fixed to not do that, or alternatively we need to rework the tests to support execution from a temporary directory…

Reply:
Essentially anything that looks like this:
```py
        test('test-compare-versions',
             test_compare_versions_sh,
             args : exe.full_path())
```
Is not testing what it seems to be testing when executed as `/tmp/autopkgtest.1bn0Ov/build.uKj/systemd/test/run-unit-tests.py `.

Reply:
Thanks for looking into this @keszybz .

So basically before the change, `/tmp/autopkgtest.1bn0Ov/build.uKj/systemd/test/run-unit-tests.py` didn't run any test because the script was looking for the tests installed in `/usr/lib/systemd/tests/`, which contains most likely nothing  ?

And now this is fixed, it appears that some tests, those actually running a binary (a generator in your case), use a hard-coded list of paths (where the binary is supposed to be installed, ie below `/usr/lib/systemd`) unless the path of the binary is passed via `$1` and in your case the generator is not installed on your system ?

Reply:
> or alternatively we need to rework the tests to support execution from a temporary directory…

The tests would need to be modified to find the binary they're testing in a parent directory, basically ?

Reply:
What about something like this ?

```
diff --git a/test/test-fstab-generator.sh b/test/test-fstab-generator.sh
index 7c060dfac7..9637bcbabb 100755
--- a/test/test-fstab-generator.sh
+++ b/test/test-fstab-generator.sh
@@ -2,14 +2,25 @@
 # SPDX-License-Identifier: LGPL-2.1-or-later
 set -e
 
+search_generator() {
+    name=system-generators/$1
+    basedir=$(realpath $(dirname "$0"))
+
+    while [ "$basedir" != '/' ] ; do
+       [ -x "$basedir/$name" ] && {
+           echo "$basedir/$name"
+           return
+       }
+       basedir=$(dirname $basedir)
+    done
+
+    return 1
+}
+
 if [[ -n "$1" ]]; then
     generator=$1
-elif [[ -x /usr/lib/systemd/system-generators/systemd-fstab-generator ]]; then
-    generator=/usr/lib/systemd/system-generators/systemd-fstab-generator
-elif [[ -x /lib/systemd/system-generators/systemd-fstab-generator ]]; then
-    generator=/lib/systemd/system-generators/systemd-fstab-generator
 else
-    exit 1
+    generator=$(search_generator systemd-fstab-generator) || exit
 fi
 
 src="$(dirname "$0")/testdata/test-fstab-generator"

```

Reply:
@keszybz @fbuihuu is this still a problem to be solved?

Reply:
https://salsa.debian.org/systemd-team/systemd/-/commit/a50fec8cf668ecdb6db8887a57e60638118e3feb changed `unit-tests` to call `/usr/lib/systemd/tests/run-unit-tests.py`. With that change, it should be all fine.